### PR TITLE
[AIRFLOW-5033] Switched to snakebite-py3

### PR DESCRIPTION
**Pull request by potiuk**
NOTE TO REVIEWER: It depends on #7278 - so please check only the last commit.

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5033

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
Snakebite is not PY3-compatible. Trying to switch to alledgedly drop-in replacement which is py3-compatible.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by potiuk** [comment]
It looks like kerberos is not working :( `pip install snakebite-py3[kerberos]` tries to compile Python2 version of kerberos libs

**Comment by zhongjiajie** [comment]
@potiuk snakebite only work for python2

**Comment by mik-laj** [comment]
It's fork of Snakebite.it should be Python 3 compatible

**Comment by sdikby** [comment]
@potiuk What is holding against dropping Snakebite and using the PyArrow interface to HDFS (https://arrow.apache.org/docs/python/filesystems.html)

**Comment by potiuk** [comment]
@sdikby nothing :) we just need someone to make PR. Happy to review it.

File: setup.py
Comment (Fokko) [inline setup.py]: In this case, we can drop the `devel_ci` altogether, right?

**Comment by Fokko** [comment]
With the community, we wanted to move to PyArrow. But this requires some more effort, see https://github.com/apache/airflow/pull/3560. I'm fine with moving to `snakebite-py3` for now since it isn't working on Python 3.

File: setup.py
Comment (potiuk) [reply setup.py]: Correct - we could!

**Comment by potiuk** [comment]
@Fokko - I am not 100% sure if it will work (hence [DO NOT MERGEI could not really test it with kerberos though - so maybe someone who uses kerberos + hdfs could test it? Do you know who could test it maybe :) ?

File: setup.py
Comment (Fokko) [reply setup.py]: I think we should drop it ;)

**Comment by Fokko** [comment]
At ING they are using hdfs+kerberos, perhaps @bolkedebruin knows someone who could give this a spin.

**Comment by potiuk** [comment]
@bolkedebruin :)? Can you help with testing that? That would be a super-easy fix for last remaining issue preventing full switch to Python 3 .

**Comment by elukey** [comment]
@potiuk hi! We met at the Apachecon, I am wondering if I can help with anything. What is the current status? From what I can see snakebyte-py3[kerberos] seems not working, maybe we could follow up with the maintainers of the package and see if we can fix? Or is there something else to do?

**Comment by potiuk** [comment]
Yeah. Snakebite-py3 is not working indeed. I think the only way to go is to implement new Hadoop operators base don another library :(

**Comment by elukey** [comment]
> Yeah. Snakebite-py3 is not working indeed. I think the only way to go is to implement new Hadoop operators base don another library :(

I opened  https://github.com/internetarchive/snakebite-py3/issues/5, the new maintainers (internet archive) are willing to get patches to make kerberos work. I'd give it a try, possibly replacing the current snakebite's kerberos dependency with something like python-gssapi (or similar)? What do you (and others) think?

**Comment by potiuk** [comment]
Sounds great @elukey if you could try it ! 

I think not having python 3 Hadoop operators is actually a blocker for 2.0 release,  so the sooner we have them the better :).  I also got in contact with people from Cloudera at the ApacheCon - @gezapeti PMC of Oozie - and maybe through Geza we can find people at Cloudera who can also be interested in making the Hadoop operator work for Airflow 2.0 and we could do it together.

**Comment by Fokko** [comment]
I would suggest moving to the pyarrow stuff: https://arrow.apache.org/docs/python/filesystems.html

The snakebyte(-py3) isn't maintained anymore.

**Comment by elukey** [comment]
> I would suggest moving to the pyarrow stuff: https://arrow.apache.org/docs/python/filesystems.html
> 
> The snakebyte(-py3) isn't maintained anymore.

I am a bit confused, earlier in this thread it was already proposed pyarrow but then it ended up being too difficult and it wasn't completed (this is my impression from reading, might be wrong). The snakebyte-py3 fork seems maintained by internet archive, it could be a good interim solution before moving to something different?

**Comment by Fokko** [comment]
Well, there's barely anything going on: https://github.com/internetarchive/snakebite-py3/commits/master

We can move to `snakebyte-py3` to remove the blocker, but in the end, I would like to migrate to PyArrow.

**Comment by elukey** [comment]
> Well, there's barely anything going on: https://github.com/internetarchive/snakebite-py3/commits/master
> 
> We can move to `snakebyte-py3` to remove the blocker, but in the end, I would like to migrate to PyArrow.

I agree, but I think that the aim of the new maintainer was just to port the existing code to py3 and then support if needed. I opened an issue and they promptly answer, so I think it is a good sign!

I am not advocating for using snakebite, only trying to find a solution to unblock the current status :)

**Comment by elukey** [comment]
Tried to work on it during the last couple of days, and hit an issue with my environment:

https://github.com/spotify/snakebite/pull/153#issuecomment-121290538

snakebite seems to rely ok `python-krbV` (the python2 only dep) only to get the kerberos principal of the running user, something that in theory can be done with other libs. The main problem seems to be that snakebite doesn't support HDFS encrypted RPCs, and fixing it might be a lot of work. I thought that it was a common use case for kerberos-enabled clusters, but apparently not?

**Comment by potiuk** [comment]
Maybe @bolkedebruin can shed some light here? I guess our only realistic option is indeed switch to a different library - I think people rely on kerberos being available for secure clusters and we cannot provide Hadoop support without it.

**Comment by elukey** [comment]
I'd love to know @bolkedebruin's opinion as well :) From my side, I can offer dev time and also test environment, since I have a kerberos with rpc encrypted cluster to use as testing environment.

**Comment by elukey** [comment]
Correcting myself: the missing feature in snakebite is encryption/decryption of the HDFS data transfer protocol, because the Hadoop RPC SASL code works fine (so basic ops with the HDFS Namenode should work). I found a bug in the new python 3 code (text vs binary string comparison ending up in always false) and now I am able to at least replicate the snakebite-py2 behavior.

**Comment by elukey** [comment]
Good news: I was able to swap krbV with gssapi, so in theory snakebite-py3 should now be able to work with Kerberos (after/if my PR is merged), no more py2 only dependencies. The only remaining thing to do is adding support for HDFS clusters with RPC encryption, that was not supported by snakebite previously (so a new feature).

Feedback/testing welcome! https://github.com/internetarchive/snakebite-py3/pull/6

**Comment by potiuk** [comment]
🎉 🎉 🎉 🎉 🎉 🎉 🎉  

I made some comments there. Hopefully this will be merged/released quickly and we can finally switch!

**Comment by elukey** [comment]
Today I have tested a simple `ls /` with pyarrow's hdfs support, and I can confirm that works withou kerberos and with kerberos+rcp-encryption. It is not a pure python solution since it uses libhdfs JNI bindings IIUC, but it seems more flexible in the long run (if we don't find more community members working on snakebite..).

**Comment by potiuk** [comment]
Great!

**Comment by elukey** [comment]
Update: my pull request for snakebite-py3 also includes support for SASL encryption for the Hadoop RPC protocol. This means that all basic commands that need only a response from the HDFS Namenode (ls / stat / etc..) work with SASL QOP (wire encryption). It was a bit hard to figure out how SASL works in this context but now it should be easier (not easy!) to also move the RPC Datanode Protocol to SASL (adding file creation etc..).

I am not getting much feedback from upstream, not sure why, I'll try to reach out to them directly to see if anybody can help.

**Comment by potiuk** [comment]
Happy to help @elukey -> please mark my github nickname when you ping them and i will also comment on that!

**Comment by elukey** [comment]
@potiuk there is a new version of snakebite-py3 on pypi! :)

**Comment by potiuk** [comment]
Yay! Is it enough to simply use the new dependency and that's it ?

**Comment by elukey** [comment]
> Yay! Is it enough to simply use the new dependency and that's it ?

Yes my understanding is that snakebite-py3[kerberos] now should work (I tested it today and it seems so, version 3.0.5). The caveat is that kerberos + wire encryption is still WIP, but I hope to add the necessary code soon. Will keep you posted :)

**Comment by potiuk** [comment]
Hey @elukey -> any news on that :)?

**Comment by elukey** [comment]
> Hey @elukey -> any news on that :)?

Hi @potiuk, unfortunately I wasn't been able to complete the code since it seems that the SASL libraries that are available (`sasl` and `pure-sasl` on pypi) are not working as I'd need. The first seems abandoned (last commit in 2016) and not working with Kerberos auth+encryption, meanwhile the latter works very well with Kerberos, but not with DIGEST-MD5 qop auth-conf (encryption, only auth is supported). The last version of `snakebite-py3` uses pure-sasl because it works really well, I opened an issue to see if the new feature can be added: https://github.com/thobbs/pure-sasl/issues/32. I don't have a lot of hope since DIGEST-MD5 auth-conf is based on ancient protocols (DES, RC4, etc..) so not really up to the current standards of AES. The Hadoop Datanode protocol seems to require it during negotiation of AES encryption (yes very ironic)..

The `sasl` lib (https://github.com/cloudera/python-sasl/tree/master/sasl) is a wrapper around the Cyrus SASL library, that should support it, but sadly I wasn't able to make it work with Kerberos (so I was blocked one step before the current issue). There is a pull request to refresh the bindings (https://github.com/cloudera/python-sasl/pull/15), but nobody from upstream is paying attention. If you know anybody from Cloudera that can help, we may have a way forward :)

Just to clarify, `snakebite-py3` now supports kerberos and works well in all the use cases, except when full encryption is requested. So currently it is doing what its python2/3 version supported, what I am trying to add is something more that was not added before. No idea if the airflow community really needs it (enough to block 2.0 I mean); if so we could work on replacing snakebite with pyarrow (even if it was tried before and I didn't get exactly what blocked the migration). Let me know :)

**Comment by Tagar** [comment]
> The sasl lib (https://github.com/cloudera/python-sasl/tree/master/sasl) is a wrapper around the Cyrus SASL library, that should support it, but sadly I wasn't able to make it work with Kerberos (so I was blocked one step before the current issue). There is a pull request to refresh the bindings (cloudera/python-sasl#15), but nobody from upstream is paying attention

@elukey 

I am author of cloudera/python-sasl#15 - closed that PR as Cloudera didn't pay attention to recent pull requests 
I've published sasl3 package on pypi
https://pypi.org/project/sasl3/
https://github.com/sparkur/python-sasl3

Please give that a try instead if `sasl` 

Thank you

**Comment by zhongjiajie** [comment]
That great, Thanks @Tagar , @elukey Maybe we should try https://github.com/sparkur/python-sasl3 in https://github.com/internetarchive/snakebite-py3 instead of `pure-sasl`

**Comment by elukey** [comment]
Thanks all for the comments, I'll surely try pure-sasl3! @potiuk can you add a comment about what I asked above? I am trying to schedule dev work on snakebite, but I didn't get if it blocks now Airflow 2.0 or not :)

**Comment by potiuk** [comment]
Sorry for the delay @elukey ! Been swamped recently :). I think as long as we have equivalent to what we had before in Airflow 1.10 (and py2) we are totally unblocked. The full-encryption might be simply a missing feature which can be implemented later.

**Comment by Tagar** [comment]
> Thanks all for the comments, I'll surely try pure-sasl3! 

@elukey I can't find pure-sasl3 .. did you mean just `sasl3` ? https://pypi.org/project/sasl3/ 

Thanks all

**Comment by elukey** [comment]
@Tagar sorry end of the day, I mixed up names :)

Speaking of pure-sasl though: https://github.com/thobbs/pure-sasl/issues/32#issuecomment-563479883 - upstream would be open to add DIGEST-md5 auth-conf support if anybody has  time/experience :) Could be an alternative road if sasl3 doesn't work for snakebite (I hope not of course). Will try to dedicate more time on snakebite during the next days!

**Comment by zhongjiajie** [comment]
No stale

**Comment by zhongjiajie** [comment]
@elukey Are you still working on this PR?

**Comment by elukey** [comment]
@zhongjiajie Hi :) As I wrote before, the blocker was removed (namely snakebite-py3 not being able to work with kerberos + python3), the only thing missing is a feature that is currently not in snakebite namely the support for RPC encryption. There are several blockers that I found while working on this, but I'll try to dedicate more time during the next days. In my opinion the issue highlighted in this task is resolved.

If the missing RPC encryption support is vital (doesn't seem from the reports of people using snakebite, but for example  I do use it in my working environment) and I can't fix the current issues the next step would be to switch to another client/library, but probably it can be done separately or as part of another pull request (maybe even after airflow 2.0?).

Let me know :)

**Comment by zhongjiajie** [comment]
@elukey  Thanks for you working here, Yeah I think it could be done in separately PR, but I think we should done before Airflow 2.0 release, cause we have roadmap to drop py2 support before Airflow 2.0 WDYT @potiuk

**Comment by potiuk** [comment]
I think we can simply proceed with this one then - I guess @elukey - can you please confirm? That it is enough to migrate to latest version of snakebite-py3 and all except the full encryption? My proposal how to proceed will be to (providing that the above is correct):

1) switch to snakebite-py3 (so basically finalise this PR) with some documentation update that full encryption is not supported

2) open a JIRA issue to implement full encryption - and hopefully someone will pick it up :)

Is that correct @elukey ?

**Comment by potiuk** [comment]
I updated the PR to latest snakebite-py3 (3.0.5+). I will wait for Travis to pass, and also I would love to get someone double-test it with the new snakebite and real HDFS.

**Comment by potiuk** [comment]
@kaxil -> I would love to cherry-pick that one before 1.10.8 if we get confirmation that it works, that would re-enable HDFS operators for 1.10.8 python 3 version :).

File: Dockerfile
Comment (ashb) [inline Dockerfile]: ?

File: setup.py
Comment (ashb) [inline setup.py:245]: Seems fine for master -- do you know if this py3 module also supports py2, or do we need to have the `;python_version>="3"` trick on 1.10.8 to support py2 and 3?

File: scripts/ci/_utils.sh
Comment (ashb) [inline scripts/ci/_utils.sh]: What caused this path change?

File: Dockerfile
Comment (mik-laj) [reply Dockerfile]: The latest Debian image from today deletes dumb-init. https://github.com/apache/airflow/pull/7278

File: scripts/ci/_utils.sh
Comment (mik-laj) [reply scripts/ci/_utils.sh]: If dumb-init is installed by apt then it appears in this directory. More info: 
https://github.com/apache/airflow/pull/5659#discussion_r372017652

File: setup.py
Comment (potiuk) [reply setup.py]: Dropped.

File: setup.py
Comment (potiuk) [reply setup.py:245]: It would be safer to use "if" in 1.10.8. I will do it when cherry-picking.

File: Dockerfile
Comment (potiuk) [reply Dockerfile]: It's only the second commit that is relevant - this change is based on #7278 indeed. Seems that latest release of python images removed pre-installed dumb-init (there was a discussion in debian devlists about removing dumb-init at some point in time). We used it now so we have to install it and use the newly installed one.

Comment (potiuk) [reply Dockerfile]: But now when I think about it - I think I will solve it differently. It's already quite some time when docker and docker-compose support --init flag which defaults to run "tini" as the init process. this will prevent us to rebuild the image and make the whole docker compose cleaner.

File: scripts/ci/_utils.sh
Comment (potiuk) [reply scripts/ci/_utils.sh]: I will replace it with the "init" field in docker compose.

File: scripts/ci/_utils.sh
Comment (ashb) [inline scripts/ci/_utils.sh]: Oh I didn't even know about this option. Nice

**Review by ashb — APPROVED** [review]

**Review by kaxil — APPROVED** [review]

**Comment by elukey** [comment]
To keep archives happy: I created https://github.com/internetarchive/snakebite-py3/issues/8 listing all the info needed to add the missing functionalities to snakebite.

**Comment by Fokko** [comment]
It looks like something is off with the CI:
```

Run mypy...................................................................................................Failed
- hook id: mypy
- duration: 1.4s
- exit code: 1
/opt/airflow/scripts/ci/in_container/run_mypy.sh: line 36: mypy: command not found
There were some mypy errors. Exiting
/opt/airflow/scripts/ci/in_container/run_mypy.sh: line 36: mypy: command not found
There were some mypy errors. Exiting
Run pylint for main sources................................................................................Failed
- hook id: pylint
- duration: 2.04s
- exit code: 1
PATH=/root:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/hive/bin:/opt/gcloud/bin
/opt/airflow/scripts/ci/in_container/run_pylint_main.sh: line 61: /usr/local/bin/pylint: No such file or directory
There were some pylint errors. Exiting
Run pylint for tests......................................................................................Skipped
- hook id: pylint-tests
Run flake8.................................................................................................Failed
- hook id: flake8
- duration: 1.47s
- exit code: 1
/opt/airflow/scripts/ci/in_container/run_flake8.sh: line 41: flake8: command not found
There were some flake8 errors. Exiting
/opt/airflow/scripts/ci/in_container/run_flake8.sh: line 41: flake8: command not found
There were some flake8 errors. Exiting
```

**Comment by potiuk** [comment]
Needs rebase

**Comment by wittfabian** [comment]
Any updates here?

**Comment by zhongjiajie** [comment]
@wittfabian No, this PR was close by stale-bot and I think no one could open it again, feel free to continue the PR by create new one if you want to and have more further idea.

**Comment by potiuk** [comment]
I will take a look shortly while I will be looking at removing conflicting requirements. The state for this snakebite is that it has been fixed by @elukey  and it should work for 2.0 with the exception of HDFS-RPC encryption. But just switching to latest released snakebite-py3 should work for all other use cases.

**Comment by zhongjiajie** [comment]
Ohh, I remember we could not reopen PR closed by stale-bot, am I wrong?

**Comment by zhongjiajie** [comment]
BTW, that a good new for AIP-3

**Comment by wittfabian** [comment]
Is it still planned to switch to PyArrow in the future? We only encounter problems with snakebite in our projects.

**Comment by zhongjiajie** [comment]
I think we should switch, but let hear Jarek opinion

**Comment by potiuk** [comment]
Yeah. If somoeone (@wittfabian ? ) would like to switch to PyArrow I am more than happy :).

**Comment by potiuk** [comment]
We even had a JIRA issue for it: https://issues.apache.org/jira/browse/AIRFLOW-2697

**Comment by wittfabian** [comment]
The last time I looked in PyArrow I had the problem that using libhdfs required Hadoop configuration (enviroment variables). Our use case is that Hadoop does not run on the same host.
Maybe I just missed something.

https://github.com/apache/arrow/blob/master/python/pyarrow/hdfs.py#L38
https://github.com/apache/arrow/blob/master/python/pyarrow/hdfs.py#L125

**Comment by elukey** [comment]
@wittfabian I am curious about what problems have you encountered with snakebite (to understand if they are big ones or small that we can solve). It is true that I didn't have the time to add support for RPC encryption, but the task seems to be big and needs time/energy (any help is needed). More info on the current blockers are listed in https://github.com/internetarchive/snakebite-py3/issues/8

Switching to PyArrow was on the table even when I started, but it was abandoned for various reasons (might not hold anymore). The main problem of snakebite is that it doesn't rely on java libs (as pyarrow does IIRC) and Hadoop still sadly uses old ciphers for SASL.

Anyway, happy if Airflow proceeds in any direction, or even more happy to collaborate with other people on snakebite-py3. Adding RPC encryption support by myself is not doable in a short amount of time.

**Comment by ashb** [comment]
Just an FYI: pyarrow is a _nightmare_ to try and install on alpline linux. I don't think this affects us, as it's an optional dep, and we provide slim Debian images.

**Comment by ashb** [comment]
@zhongjiajie as committers we can reopen stalebot closed PRs, and we can also add the "pinned" label to prevent stalebot closing them in the first place.

**Comment by Tagar** [comment]
> Just an FYI: pyarrow is a _nightmare_ to try and install on alpline linux. I don't think this affects us, as it's an optional dep, and we provide slim Debian images.

I had issues with pyarrow dependency issues too. Boost, a popular C++ library was most often caused a lot of those issues. Arrow 0.16 release in Jan made Boost optional. 
https://arrow.apache.org/blog/2020/02/12/0.16.0-release/ 

> Notably, Boost is no longer required. We invested effort to vendor some small essential dependencies: Flatbuffers, double-conversion, and uriparser. Many optional features requiring external libraries, like compression and GLog integration, are now disabled by default. Several subcomponents of the C++ project like the filesystem API, CSV, compute, dataset and JSON layers, as well as command-line utilities, are now disabled by default

So it might be a good idea to try latest pyarrow now. 

On a separate note, I sent a comment above 

> I've published sasl3 package on pypi
> https://pypi.org/project/sasl3/
> https://github.com/sparkur/python-sasl3
> 
> Please give that a try instead if sasl 
> 
> 

That should address the issue described by @elukey in this comment 
https://github.com/apache/airflow/pull/5659#issuecomment-559814094

**Comment by wittfabian** [comment]
@elukey I'm not 100 percent sure, but we had a lot of problems with the installation. Maybe some problems have been solved with new versions. I looked at some details again yesterday, it is possible that some of the problems also come from kerberos.

I can live with any solution that runs on Python 3.

**Comment by zhongjiajie** [comment]
> @zhongjiajie as committers we can reopen stalebot closed PRs, and we can also add the "pinned" label to prevent stalebot closing them in the first place.

Thanks @ashb  for the clarification

**Comment by elukey** [comment]
> On a separate note, I sent a comment above
> 
> > I've published sasl3 package on pypi
> > https://pypi.org/project/sasl3/
> > https://github.com/sparkur/python-sasl3
> > Please give that a try instead if sasl
> 
> That should address the issue described by @elukey in this comment
> [#5659 (comment)](https://github.com/apache/airflow/pull/5659#issuecomment-559814094)

@Tagar I tried today sasl3 but I get to the same error that I had with sasl, namely the following error message on the Namenode when trying to send a getFileInfo RPC call:

```
javax.security.sasl.SaslException: Problems unwrapping SASL buffer [Caused by GSSException: Defective token detected (Mechanism level: Kerberos GSS-API Mechanism Token:Defective Token ID!)]
```
I recall that at the time I solved the problem simply switching to pure-sasl, I'll try to follow up and see if there is something new that I can debug with fresh eyes.

**Comment by elukey** [comment]
Good news: I got unblocked, I was able to make snakebite working with `sasl3` after a lot of time spent between swearing and crying :D, details in https://github.com/internetarchive/snakebite-py3/issues/8#issuecomment-625748266

The next step is to implement the missing bits in snakebite, hoping that I'll not find any pitfall like this one. Fingers crossed!

**Comment by Fokko** [comment]
@elukey Can you create a PR, would be great to get it into Airflow :)

**Comment by elukey** [comment]
> @elukey Can you create a PR, would be great to get it into Airflow :)

I am currently trying to solve the last missing bit (namely encrypted RPCs between client and Datanodes, I'll file a PR to the main snakebite-py3 repo as soon as I'll have something working and tested. Just as FYI, the current version of snakebite-py3 supports py3, non-encrypted-rpcs and a subset of encrypted RPCs (the ones between client and Namenode). The work that I am doing now is to have a complete support, but 75% of the work is already upstreamed and working :)

**Comment by elukey** [comment]
> > Just an FYI: pyarrow is a _nightmare_ to try and install on alpline linux. I don't think this affects us, as it's an optional dep, and we provide slim Debian images.
> 
> I had issues with pyarrow dependency issues too. Boost, a popular C++ library was most often caused a lot of those issues. Arrow 0.16 release in Jan made Boost optional.
> https://arrow.apache.org/blog/2020/02/12/0.16.0-release/
> 
> > Notably, Boost is no longer required. We invested effort to vendor some small essential dependencies: Flatbuffers, double-conversion, and uriparser. Many optional features requiring external libraries, like compression and GLog integration, are now disabled by default. Several subcomponents of the C++ project like the filesystem API, CSV, compute, dataset and JSON layers, as well as command-line utilities, are now disabled by default
> 
> So it might be a good idea to try latest pyarrow now.

@potiuk @Fokko due to the complexity of adding encryption support to snakebite (at least for me), I would in parallel try again pyarrow and see if it suits the Airflow needs. I am doing some hacking now and I am learning some things, but it would be better to explore other paths as well. What do you think?

**Comment by potiuk** [comment]
All for it. I already installed pyarrow few days ago as it was needed by apache beam provider!

**Comment by potiuk** [comment]
Closing this one as we already have snakebite-py3 in master airflow. I think we might need separate issue  for PyArrow implementation.

**Comment by elukey** [comment]
To keep archives happy: I tried hard to make snakebite compatible with RPC encryption, but found a lot of road blockers, all explained in https://github.com/internetarchive/snakebite-py3/issues/8#issuecomment-651769727. For the moment I am inclined to stop working on this and let Airflow/others to experiment with pyarrow, it seems a more promising strategy to follow :)

**Pull request closed without merging**
